File: magellan-legacy/src/main/java/com/wealthfront/magellan/EvictingQueue.java
package com.wealthfront.magellan;

import java.util.ArrayDeque;
import java.util.Deque;
import java.util.Iterator;

class EvictingQueue<E> implements Iterable<E> {

  private final Deque<E> queue;
  private final int maxSize;

  EvictingQueue(final int maxSize) {
    this.maxSize = maxSize;
    queue = new ArrayDeque<>(maxSize);
  }

  public void add(E e) {
    if (queue.size() == maxSize) {
      queue.remove();
    }
    queue.add(e);
  }

  @Override
  public Iterator<E> iterator() {
    return queue.iterator();
  }

}

File: magellan-kotlin/src/main/java/com/wealthfront/magellan/kotlin/EvictingQueue.kt
package com.wealthfront.magellan.kotlin

import java.util.*

internal class EvictingQueue<E>(private val maxSize: Int) : Iterable<E> {

  private val queue: Deque<E>

  init {
    queue = ArrayDeque(maxSize)
  }

  fun add(e: E) {
    if (queue.size == maxSize) {
      queue.remove()
    }
    queue.add(e)
  }

  override fun iterator(): Iterator<E> = queue.iterator()

}
